src/main/kotlin/com/github/pemistahl/lingua/internal/TrainingDataLanguageModel.kt

```diff
@@ -147,4 +145,94 @@ internal data class TrainingDataLanguageModel(
             return ngramProbabilities
         }
     }
+
+    internal class RelativeFrequencies private constructor(private val data: Map<Long, Entries>) {
+
+        operator fun get(ngram: String): Float = data[computeHighHash(ngram)]?.get(ngram) ?: 0F
+
+        private class Entries(private val chars: ByteArray, private val frequencies: FloatArray) {
+
+            val size get() = frequencies.size
+
+            operator fun get(ngram: String): Float {
+                var low = 0
+                var high = size - 1
+                while (low <= high) {
+                    if (low + 8 < high) {
+                        // bisection search
+                        val middle = (low + high) / 2
+                        val diff = compareNgram(middle, ngram)
+                        if (diff < 0) low = middle + 1
+                        else if (diff > 0) high = middle - 1
+                        else return frequencies[middle]
+                    } else {
+                        // linear search
+                        for (i in low..high) {
+                            if (compareNgram(i, ngram) == 0) return frequencies[i]
+                            return 0F
+                        }
+                    }
+                }
+                return 0F
+            }
+
+            /**
+             * Compare lower bits only.
+             */
+            private fun compareNgram(pos: Int, ngram: String): Int {
+                val base = pos * ngram.length
+                repeat(ngram.length) { i ->
+                    val diff = chars[base + i].compareTo(ngram[i].code.and(0xFF))
+                    if (diff != 0) return diff
+                }
+                return 0
+            }
+        }
+
+        companion object {
+
+            /**
+             * Compare low bits of each character.
+             * String length must be the same.
+             */
+            private object LowByteComparator : Comparator<String> {
+                override fun compare(o1: String, o2: String): Int {
+                    for (i in o1.indices) {
+                        val res = o1[i].code.and(0xFF) - o2[i].code.and(0XFF)
```

Comment:
Should this for consistency be a lower case `x`?
```suggestion
                        val res = o1[i].code.and(0xFF) - o2[i].code.and(0xFF)
```